### PR TITLE
Add Bitbucket Pipelines known host module

### PR DESCRIPTION
**Pull request by catcombo**
##### SUMMARY

Manages Bitbucket Pipelines [known hosts](https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/pipelines_config/ssh/known_hosts).

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

bitbucket_pipeline_known_host

### CODE REVIEW

File: lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community',
+}
+
+DOCUMENTATION = r'''
+---
+module: bitbucket_pipeline_known_host
+short_description: Manages Bitbucket pipelines known hosts
```
Comment (zyv) [inline lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py]:
```suggestion
short_description: Manages Bitbucket pipeline known hosts
```

File: lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+description:
+  - Create or delete a repository level known hosts.
```
Comment (zyv) [inline lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py]:
```suggestion
  - Manages Bitbucket pipeline known hosts under the "SSH Keys" menu.
```

File: lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+      - If not set the environment variable C(BITBUCKET_CLIENT_ID) will be used.
+    type: str
+  client_secret:
+    description:
+      - The OAuth consumer secret.
+      - If not set the environment variable C(BITBUCKET_CLIENT_SECRET) will be used.
+    type: str
+  repository:
+    description:
+      - Repository name.
```
Comment (zyv) [inline lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py]:
```suggestion
      - The repository name.
```

File: lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    type: str
+    required: true
+  username:
+    description:
+      - Repository owner.
```
Comment (zyv) [inline lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py]:
```suggestion
      - The repository owner.
```

File: lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+    type: str
+    required: true
+  name:
+    description:
+      - Hostname of the known host.
```
Comment (zyv) [inline lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py]:
```suggestion
      - The FQDN of the known host.
```

File: lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+    type: str
+    required: true
+  key:
+    description:
+      - Public key.
```
Comment (zyv) [inline lib/ansible/modules/source_control/bitbucket/bitbucket_pipeline_known_host.py]:
```suggestion
      - The public key.
```

**Review by zyv — CHANGES_REQUESTED** [review]
Doc fixes, please integrate using the GUI.

**Comment by catcombo** [comment]
@ansibot ready_for_review

**Review by zyv — APPROVED** [review]
LGTM!

**Comment by zyv** [comment]
While testing this and other modules that we are trying to open source on our infra, we actually came up with a design question for Ansible developers.

We have following modules that operate on individual Bitbucket Pipeline configuration entities:

* `bitbucket_pipeline_known_host`
* `bitbucket_pipeline_variable` (already merged in #54049)
* `bitbucket_access_key` (replacement for `bitbucket_deploy_key`)

Arguably, a better design would be to operate on complete setting blocks (lists of pipeline variables, list of known hosts). One might argue the same for access keys, but probably there the argument is more thin. I personally would prefer to manage those one by one, especially since there is mostly just one.

The advantages of operating on lists of entities would be that:

1. The modules will be much faster, especially for large number of variables and known hosts; currently our playbooks are taking minutes to run for dozens of repositories with dozens of variables.

2. One can implement (in a second step) a nice diff-mode like in Terraform, and users will be sure that there will be no unmanaged variables and/or known hosts on their repositories.

We can't really think of strong arguments against this, other than that it's apparently unusual for Ansible modules to behave this way. However, there are precedents like `ec2_vpc_route_table` module or `ec2_vpc_nacl` module. Supposedly, the justification is the same - they would be extremely slow if one would have to set up route by route or ACL by ACL via `with_items`...

Any thoughts on this? How should we proceed? If you think it's a much better idea, then we will rework the modules quickly. Thanks!

/cc @dagwieers @gundalow @Akasurde

**Comment by dagwieers** [comment]
Do you mean something like this?
https://github.com/ansible/proposals/issues/71

I tried to get something like this approved but the consensus was to solve this using roles. (I don't think you can properly solve this using roles, as it is slow and painful).

**Comment by dagwieers** [comment]
The actual proposal is at: https://github.com/ansible/proposals/issues/71#issuecomment-417823437

**Comment by zyv** [comment]
@dagwieers awesome, yes, this is exactly what I was thinking about! I agree that solving it on core level is much better than trying to come up with a specific solution for each individual module, i.e.

```
pipeline_variables:
  - { name: AWS_ACCESS_KEY, value: ABCD1234 }
  - { name: AWS_SECRET, value: qwe789poi123vbn0, secured: True }

- name: Create or update pipeline variables from the list
  bitbucket_pipeline_variable:
    repository: 'bitbucket-repo'
    username: bitbucket_username
    variables: '{{ pipeline_variables }}'
    state: present	
```

So I guess the bottom line for us would be that we keep our current design, and implement `dataset` support when the proposal is implemented...

Is there anything else we could/should do to push our PRs forward :) ?

**Comment by dagwieers** [comment]
@zyv Feel free to show your support in the proposal ;-)

**Review by dagwieers — APPROVED** [review]
LGTM, and includes unit tests !

**Pull request merged**